### PR TITLE
Implement a new option that adds support for copying shadow root content

### LINKED ISSUE #176: React component wrapped inside custom element (web component)
Thank you for react-to-print. Excellent job! In my application users can toggle between a dark and a light (Bootstrap-based) theme. For when the dark theme is active, I figure I "wrap" my React component inside a custom element (web component) to take advantage of the shadow DOM. That would allow me to switch to light theme just for printing purposes without affecting the dark theme that the user is looking at.

When I use react-to-print to print this React class component directly it works as advertised. When I wrap it inside a custom element, everything seem to go through the motions but nothing is rendered. The react-to-print callbacks are called and the React component's render method is called. When using the custom element directly inside another React component (no printing), it renders to the screen just fine.

Would you have any idea what might be going on here?

### PR DESCRIPTION
**Pull request by boehlke**
Resolves #176

### CODE REVIEW

File: src/hooks/useReactToPrint.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { useCallback } from "react";
 
-import { Font } from "../types/Font";
+import { Font } from "../types/font";
```
Comment (MatthewHerbst) [inline src/hooks/useReactToPrint.ts:3]: Is this change intentional?

File: src/types/UseReactToPrintOptions.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { RefObject } from "react";
-import type { Font } from "./Font";
+import type { Font } from "./font";
```
Comment (MatthewHerbst) [inline src/types/UseReactToPrintOptions.ts:2]: Is this change intentional?

File: src/types/UseReactToPrintOptions.ts [hunk 3]
```diff
@@ -54,4 +54,6 @@ export interface UseReactToPrintOptions {
     print?: (target: HTMLIFrameElement) => Promise<any>;
     /** When passed, prevents `console` logging of errors */
     suppressErrors?: boolean;
-}
\ No newline at end of file
+    /** When passed, shadow root content will be copied */
+    copyShadowRoots?: boolean;
```
Comment (MatthewHerbst) [inline src/types/UseReactToPrintOptions.ts:59]: Can we add this to the README as well please?

File: src/utils/clone.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+export function cloneShadowRoots(sourceNode: Node, targetNode: Node) {
+    if(sourceNode instanceof Element && targetNode instanceof Element) {
```
Comment (MatthewHerbst) [inline src/utils/clone.ts]:
```suggestion
    if (sourceNode instanceof Element && targetNode instanceof Element) {
```

File: src/utils/handlePrintWindowOnLoad.ts [hunk 5]
```diff
@@ -1,7 +1,8 @@
 import { logMessages } from "./logMessage";
 import { startPrint } from "./startPrint";
-import { Font } from "../types/Font";
+import { Font } from "../types/font";
```
Comment (MatthewHerbst) [inline src/utils/handlePrintWindowOnLoad.ts]: Is this change intentioinal?

File: src/utils/handlePrintWindowOnLoad.ts [hunk 6]
```diff
@@ -61,7 +62,10 @@ export function handlePrintWindowOnLoad(
     const domDoc = printWindow.contentDocument || printWindow.contentWindow?.document;
 
     if (domDoc) {
-        domDoc.body.appendChild(clonedContentNode);
+        const appendedContentNode = domDoc.body.appendChild(clonedContentNode);
+        if(options.copyShadowRoots) {
```
Comment (MatthewHerbst) [inline src/utils/handlePrintWindowOnLoad.ts]:
```suggestion
        if (options.copyShadowRoots) {
```

Comment (MatthewHerbst) [reply src/utils/handlePrintWindowOnLoad.ts]: Nit: can you please destructure `copyShadowRoots` above like on lines 49-56

File: src/utils/clone.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        const shadowRoot = sourceNode.shadowRoot;
+        if (shadowRoot !== null) {
+            // Create a new Shadow Root for the target element
+            const copiedShadowRoot = targetNode.attachShadow({mode: shadowRoot.mode});
```
Comment (MatthewHerbst) [inline src/utils/clone.ts]:
```suggestion
            const copiedShadowRoot = targetNode.attachShadow({ mode: shadowRoot.mode. });
```

File: src/utils/clone.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+            // Clone the contents of the original ShadowRoot manually
+            const originalShadowContent = shadowRoot.innerHTML;
+
+            // Create a temporary container to hold the content to copy
+            const tempContainer = targetNode.ownerDocument.createElement(shadowRoot.children[0].tagName);
```
Comment (MatthewHerbst) [inline src/utils/clone.ts]: Will the shadow root always have at least one child?

File: src/utils/clone.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            tempContainer.innerHTML = originalShadowContent;
+
+            // Clone and append the Shadow DOM content
+            copiedShadowRoot.appendChild(tempContainer);
+
+            // Recursively clone any nested Shadow DOMs within this Shadow DOM content
```
Comment (MatthewHerbst) [inline src/utils/clone.ts]: This recursion and the one below is what scared me from doing this in the past, I'm worried about performance problems on pages with large amounts of shadow content. Maybe we worry about that if/when it happens though

**Review by MatthewHerbst — COMMENTED** [review]
This looks great, thanks! Just a few things to cleanup please, mostly we need to add the docs for it to the README

File: src/types/UseReactToPrintOptions.ts (hunk 2 above)
Comment (boehlke) [reply src/types/UseReactToPrintOptions.ts:2]: There is only a "font.ts", no "Font.ts". My build failed with 'from "./Font"...

File: src/utils/handlePrintWindowOnLoad.ts (hunk 5 above)
Comment (boehlke) [reply src/utils/handlePrintWindowOnLoad.ts]: Yes... like pointed out above.

File: src/utils/clone.ts (hunk 9 above)
Comment (boehlke) [reply src/utils/clone.ts]: What do think about a warning in the README to address this issue.
I optimized/simplified the clone routine, treewalker should be quite efficient. Please have a look.

File: src/utils/clone.ts (hunk 8 above)
Comment (boehlke) [reply src/utils/clone.ts]: Simplified that... the old code version was a bit hacky. Much cleaner now.

File: src/hooks/useReactToPrint.ts (hunk 1 above)
Comment (boehlke) [reply src/hooks/useReactToPrint.ts:3]: Can you resolve this if lower case works fine for you as well? Do you work with NTFS? For case sensitive FS, this should lead to an error.

File: src/types/UseReactToPrintOptions.ts (hunk 2 above)
Comment (MatthewHerbst) [reply src/types/UseReactToPrintOptions.ts:2]: I think my computer is lying to me!

Screenshot 2024-08-19 at 1 51 58 AM
Screenshot 2024-08-19 at 1 52 09 AM

File: src/utils/clone.ts (hunk 9 above)
Comment (MatthewHerbst) [reply src/utils/clone.ts]: I like that, thanks!

File: src/utils/clone.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    const sourceElements = collectElements(sourceNode as HTMLElement);
+    const targetElements = collectElements(targetNode as HTMLElement);
+
+    for (let i = 0; i < sourceElements.length; i++) {
+        const sourceElement = sourceElements[i];
+        const targetElement = targetElements[i];
+
+        const shadowRoot = sourceElement.shadowRoot;
+        if (shadowRoot !== null) {
```
Comment (MatthewHerbst) [inline src/utils/clone.ts:34]: Should we guard that `targetElement` is also defined, since it's possible `targetElements` may have a different length than `sourceElements` (this would likely only happen if the user isn't waiting for their DOM to be settled before printing, but if they make a mistake there this will hard crash them. I just added a similar check when iterating over `canvas` elements): https://github.com/MatthewHerbst/react-to-print/blob/8188f8406ca17752a513b5bf46172cf953df3805/src/utils/handlePrintWindowOnLoad.ts#L121-L127

File: src/utils/clone.ts (hunk 8 above)
Comment (MatthewHerbst) [reply src/utils/clone.ts]: `TreeWalker` is nice!

File: src/utils/clone.ts (hunk 10 above)
Comment (boehlke) [reply src/utils/clone.ts:34]: Please review, I added a similar error handling like you suggested.

Comment (MatthewHerbst) [reply src/utils/clone.ts:34]: Perfect, thanks!

**Review by MatthewHerbst — APPROVED** [review]
Thanks so much for this! I'll get it into a `beta-3` release this evening 🥳

**Comment by MatthewHerbst** [comment]
Published in `v3.0.0-beta-3` 🥳
